drivers/mrf24j40: Add mrf24j40 radio support

bergzand opened the pull request:
This patch collection adds support for the mrf24j40 radio from microchip.

supported:
- transmit/receive
- most of netdev2 get/set (ack_req, csma setting...)
- sleep/reset

not supported:
- retransmission settings (not supported by the radio)
- off state (not supported by the radio)
- preloading (hangs the driver, did not investigate any further)

As visible in the commits, this PR builds on the initial work of @Nenaco, but the whole driver logic is rewritten to prevent a deadlock. I hope I've fixed most of the comments from the original PR ( #5792).

OlegHahm commented:
Just a site note: apparently the mail address you're using for your git commits seems not to be added to your GitHub profile.

bergzand commented:
Thanks for the notice, should be fixed now

bergzand commented:
Whoops, something went wrong there

bergzand commented:
And some git stupidity of my own.

Driver should be uncrustified and static check gives green on everything except a PR squash message

OlegHahm commented:
@PeterKietzmann, you have the hardware in Hamburg, right? Any candidate for review and testing?

PeterKietzmann commented:
Yes I have and Hauke should as well. The candidate is me but I can't promise to have an in-detail review and test until next Friday (feature freeze)

Nenaco commented:
Hi Folks,
thanks for this new PR on the topic MRF24J40 driver development. I'm very sad that I had no time to drive the PR #5792 further on, but there was so much pressure on my person so that my engagement stopped for some weeks. I have only seen in the news letter, that there's something going on about the MRF24J40 in the background. Great!
Regards, 
Neo

haukepetersen commented:
Yes, I also do have 2 of these somewhere in my drawer. So do I get this right, that this PR is an alternative (or even successor) of #5792?

haukepetersen commented on drivers/Makefile.dep:
```diff
@@ -18,6 +18,19 @@ ifneq (,$(filter at86rf2%,$(USEMODULE)))
 endif
 
+ifneq (,$(filter mrf24j40,$(USEMODULE)))
+  USEMODULE += mrf24j40
```
this line strikes me as quite useless...

bergzand replied:
I think that one is directly copied from the at86rf2% above it. While it makes sense there, it looks indeed useless here.

PeterKietzmann commented:
It's a successor. Wanted to give it a quick try tomorrow but as this needs adoptions to the SPI rework I assume it might be too late for this release. But the next hack'n'ack is coming and this driver has been on my todo list for so long.

haukepetersen commented on drivers/include/mrf24j40.h:
```diff
@@ -0,0 +1,455 @@
+/**
+ * @brief   Maximum possible packet size in byte
+ */
+#define MRF24J40_MAX_PKT_LENGTH        (IEEE802154_FRAME_LEN_MAX)
```
Why redefine `IEEE802154_FRAME_LEN_MAX` here?

haukepetersen commented on drivers/include/mrf24j40.h:
```diff
@@ -0,0 +1,455 @@
+ */
+#define MRF24J40_DEFAULT_ADDR_SHORT    (0x0230)
+#define MRF24J40_DEFAULT_ADDR_LONG     (0x1222334455667788)
+/** @} */
```
Please use the newly (to be) introduced `uuid` module for address generation -> #6314

haukepetersen commented on drivers/include/mrf24j40.h:
```diff
@@ -0,0 +1,455 @@
+#define MRF24J40_MIN_CHANNEL           (IEEE802154_CHANNEL_MIN)
+#define MRF24J40_MAX_CHANNEL           (IEEE802154_CHANNEL_MAX)
+#define MRF24J40_DEFAULT_CHANNEL       (IEEE802154_DEFAULT_CHANNEL)
+/** @} */
```
again, IMHO no need to redefine, just use the IEEE802154 defines directly

haukepetersen commented on drivers/mrf24j40/include/mrf24j40_params.h:
```diff
@@ -0,0 +1,73 @@
+#endif
+
+#define MRF24J40_PARAMS_DEFAULT    { .spi = MRF24J40_PARAM_SPI, \
+                                     .spi_speed = MRF24J40_PARAM_SPI_SPEED, \
```
indention one off

haukepetersen commented on tests/driver_mrf24j40/Makefile:
```diff
@@ -0,0 +1,16 @@
+APPLICATION = driver_mrf24j40
```
Though I appreciate the effort put into this test application, I tend to say we can savely drop it. Testing the driver would simply be:
- build the `examples/gnrc_networking` with `USEMODULE=mrf24j40 BOARD=BLBLBLBL make`
- set/get options (addresses, channel, pan)
- send packets
- done

Or did I miss any additional functionality provided by this test app?

phectori replied:
It is basically the same driver test as the `driver_at86rf2xx` test (so, not much effort). Since (I believe) the `gnrc_networking` example also has a test on the link layer (the function called `txtsnd`)  I agree.

miri64 replied:
> Or did I miss any additional functionality provided by this test app?

Automated testing?

> Since (I believe) the gnrc_networking example also has a test on the link layer (the function called txtsnd) I agree.

#6397 removed that and fixed #5034 through that.

miri64 replied:
(also, if all other drivers are able to send via IP and not via `mrf24j40` it is safe to assume that device is at fault ;-))

bergzand replied:
So this test should be removed then?

PeterKietzmann replied:
Yes. Please remove it. One can use `gnrc_networking` or `default`. Later #5286 will unify these tests anyway.

bergzand commented:
I fixed the first set of issues.
I can rework the driver to use the new uuid framework and the new SPI api if you really want this driver in the new release. I should have time to rebase it on those patches this evening or tomorrow morning (UTC+1). 
Otherwise I'm going to wait until those are merged before incorporating those changes in the driver

PeterKietzmann commented:
@bergzand that would be awesome in any way! But to have a realistic idea if this would make it into the release I will try to connect one until the afternoon

PeterKietzmann commented:
@bergzand sorry I didn't make it yesterday but now I gave it a shot and I was pretty happy to see the state of the driver. I used `gnrc_networking` and "palyed around" a bit. Setting parameters (number of retransmissions is not adjustable?) as well as hardcore pinging of small and big fragmented IPv6 packets between this device and an Atmel with at86rf2xx radio. Now we can eiter **hurry up very much** and try to get this in until the afternoon to let this driver be part of the 2016.01 release. But the uuid is already merged and the SPI PR has priority. Or we can give it just **a bit** more time and handle this calmly.

bergzand commented:
@PeterKietzmann Thank you for testing the driver.
> @bergzand sorry I didn't make it yesterday but now I gave it a shot and I was pretty happy to see the state of the driver. I used gnrc_networking and "palyed around" a bit. Setting parameters (number of retransmissions is not adjustable?) as well as hardcore pinging of small and big fragmented IPv6 packets between this device and an Atmel with at86rf2xx radio.

The number of retransmissions is unfortunately not configurable. Blame Microchip for that one.
> Now we can eiter hurry up very much and try to get this in until the afternoon to let this driver be part of the 2016.01 release. But the uuid is already merged and the SPI PR has priority. Or we can give it just a bit more time and handle this calmly.

I just rebased the driver and added another commit applying the uuid implementations. Besides my node having a different (static) IPv6 address now, everything seems to work.

I should have some time today to apply the SPI API change, but your call if you want to hurry things up or test a bit more before merging

bergzand commented:
Test removed

PeterKietzmann commented on drivers/mrf24j40/mrf24j40.c:
```diff
@@ -0,0 +1,173 @@
+ * @brief       Implementation of public functions for MRF24J40 drivers
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ * @author      <neo@nenaco.de>
```
Please add alias here Neo Nenaco

PeterKietzmann commented on drivers/mrf24j40/mrf24j40.c:
```diff
@@ -0,0 +1,173 @@
+    mrf24j40_set_chan(dev, IEEE802154_DEFAULT_CHANNEL);
+
+    /* configure Immediate Sleep and Wake-Up mode */
+    mrf24j40_reg_write_short(dev, MRF24J40_REG_WAKECON, MRF24J40_WAKECON_IMMWAKE);            /* enable Immediate Wake-up mode */
```
Pleas put the comment somewhere else to make the line shorter

PeterKietzmann commented on drivers/mrf24j40/mrf24j40.c:
```diff
@@ -0,0 +1,173 @@
+
+void mrf24j40_tx_prepare(mrf24j40_t *dev)
+{
+    /* check */
```
Needed?

PeterKietzmann commented on drivers/mrf24j40/mrf24j40.c:
```diff
@@ -0,0 +1,173 @@
+void mrf24j40_tx_prepare(mrf24j40_t *dev)
+{
+    /* check */
+    DEBUG("[mrf24j40] TX_Prepare, Current: %x\n", dev->state);
```
Current state?

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -0,0 +1,499 @@
+        case 26:    channel_value = 0xf3;
+            break;
+        default:
+            channel_value = 0x3;
```
I would prefer to do nothing in case channel is not available.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -0,0 +1,499 @@
+    /* get current register settings */
+    tmp  = mrf24j40_reg_read_short(dev, MRF24J40_REG_TXMCR);
+    /* clear csma bits */
+    tmp  &= ~(0x7);
```
Needed? See line 304

bergzand replied:
Yes, Line 304 does an `OR` operation with the new settings so that the other settings in that register are not affected. The clear is needed so that an old CSMA setting does not affect the new setting. For example, `0b100` is for 4 backoffs. Changing to 3 (`0b011`) backoffs without this clear would result in accidentaly setting `111` for the CSMA bits in the register.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -0,0 +1,499 @@
+    mrf24j40_reg_write_short(dev, MRF24J40_REG_TXMCR, tmp);
+}
+
+void mrf24j40_set_csma_backoff_exp(mrf24j40_t *dev, uint8_t min, uint8_t max)
```
Please remove of implement

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -0,0 +1,499 @@
+                 * in the CSMA-CA algorithm. Note that if this value is set
+                 * to �0�, collision avoidance is disabled. */
+                mrf24j40_reg_write_short(dev, MRF24J40_REG_TXMCR, tmp);
+
```
Please remove blank lines here and line 380 and 413 and 420

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -0,0 +1,499 @@
+        /* Wake mrf up */
+        mrf24j40_reg_write_short(dev, MRF24J40_REG_WAKECON, MRF24J40_WAKECON_IMMWAKE | MRF24J40_WAKECON_REGWAKE);
+        /* undocumented delay, needed for stable wakeup */
+        xtimer_usleep(50);
```
Hmmm. Who added this? And how was it observed? Did anyone compare to the linux driver?

bergzand replied:
Ah, yes, I remember that one :).
This is advice from the [microchip forums](http://www.microchip.com/forums/m581653.aspx) after I could not get the mrf24j40 to wake up based on the instructions from the datasheet.

What I think what happens is that when the device is put into sleep mode, it switches to the 32kHz external oscillator (or 100kHz internal depending on your configuration). Now, the datasheet says: 
  Wake-up can be performed in one of two methods:
  1. Wake-up on WAKE pin 15. To enable the WAKE
  pin, set the WAKEPAD (0x0D<5>) bit to ‘1’ and
  set the WAKE pin polarity. Set the WAKEPOL
  (0x0D<7>) bit to ‘1’ for active-high signal, or
  clear to ‘0’ for active-low signal.
  ...

I think what happens is that the toggle of that bit has to be sampled by the low speed sleep clock. So: set bit high -> wait for bit to be detected by the mrf24j40 -> set bit low again. I have experienced that toggling the bit without a wait, the mrf24j40 will not wake up and the driver gets stuck waiting for the state machine reset. This only seems to happen when using the spi register wake up instead of the wake pin. Since this method does not cause a lot of additional delays additional to the 2ms wake delay, I think it's worth the pin it saves.

I don't know if the linux driver also has this delay, but if it uses the wake pin, it's probably not needed.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_internal.c:
```diff
@@ -0,0 +1,244 @@
+
+    /* mrf24j40_set_interrupts */
+    mrf24j40_reg_write_short(dev, MRF24J40_REG_INTCON, ~(MRF24J40_INTCON_RXIE | MRF24J40_INTCON_TXNIE));
+    //Wait until the RFSTATE machine indicates RX state
```
Maybe block until RX state achieved? Please don't use `//` for comments

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_internal.c:
```diff
@@ -0,0 +1,244 @@
+    uint16_t i;
+
+    for (i = 0; i < len; i++) {
+        mrf24j40_reg_write_long(dev, i, data[i]);
```
Long words to fill the fifo?

bergzand replied:
No, the mrf24j40 has two types of control registers, short (with 8 bit addresses) and long (16 bit addresses) as the datasheet calls them. The fifo's are in the "long" address space, thus a reg_write_long is needed

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_internal.c:
```diff
@@ -0,0 +1,244 @@
+
+void mrf24j40_update_tasks(mrf24j40_t *dev)
+{
+
```
remove blank line

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+/*
+ * Copyright (C) 2016 Hochschule für Angewandte Wissenschaften Hamburg
```
Copyright and author don't correlate

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+    memset(&netdev->stats, 0, sizeof(netstats_t));
+#endif
+    /* reset device to default values and put it into RX state */
+    mrf24j40_reset(dev);                                            /* is done later */
```
Is done later?

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+     * device from overwriting the buffer while we're reading it.
+     * BBREG1 only contains the RXDECINV setting, so no overwriting
+     * other bits with this.
+     */
```
"Disable receiving while reading the RX fifo"?

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+    int res;
+    switch (opt) {
+        case NETOPT_CHANNEL_PAGE:
+            if (max_len < sizeof(uint16_t)) {
```
We should consider to replace this with `asserts` as done in #4419. But no worries. That can eb done in a follow up PR.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+{
+    mrf24j40_t *dev = (mrf24j40_t *) netdev;
+
+#ifdef MODULE_NETSTATS_L2
```
superfluous

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+                netdev->event_callback(netdev, NETDEV2_EVENT_TX_COMPLETE);
+            }
+        }
+#endif
```
I assume the whole block from 546:566 is **not** only relevant for netstats?

bergzand replied:
I assumed it was irrelevant since in `sys/net/gnrc/link_layer/netdev2/gnrc_netdev2.c` the switch statement also has the same ifdef around the handlers of those callback options (sys/net/gnrc/link_layer/netdev2/gnrc_netdev2.c 76:83. If you want the ifdef removed or minimized, I'm okay with that, but as it is now, it saves an SPI transfer when L2 stats are not enabled and I would like to keep that

PeterKietzmann replied:
@bergzand thanks for the explanation. Leave it as is for now. Probably I will come back to you on that point.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+            netdev->event_callback(netdev, NETDEV2_EVENT_RX_COMPLETE);
+        }
+        dev->pending &= ~(MRF24J40_TASK_RX_READY);
+    } // end of RXIF check
```
Comment style

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,578 @@
+#endif
+
+    }
+    /* RECEIVE INTERRUPT OCCURED */
```
why capitaly?

PeterKietzmann commented on sys/auto_init/netif/auto_init_mrf24j40.c:
```diff
@@ -0,0 +1,72 @@
+ * @file
+ * @brief   Auto initialization for mrf24j40 network interfaces
+ *
+ * @author  <neo@nenaco.de>
```
Copyright and author?

PeterKietzmann commented on sys/auto_init/netif/auto_init_mrf24j40.c:
```diff
@@ -0,0 +1,72 @@
+ * @{
+ */
+#define MRF24J40_MAC_STACKSIZE     (THREAD_STACKSIZE_DEFAULT)
+#define MRF24J40_MAC_PRIO          (THREAD_PRIORITY_MAIN - 4)
```
please use
```
#ifndef MRF24J40_MAC_PRIO
#define MRF24J40_MAC_PRIO          (GNRC_NETDEV2_MAC_PRIO)
#endif
```

bergzand commented on drivers/mrf24j40/include/mrf24j40_params.h:
```diff
@@ -0,0 +1,73 @@
+ * @{
+ * @file
+ * @brief       Default configuration for the MRF24J40 driver
+ *
```
Is fixed in the last commit, but github doesn't detect it as outdated

bergzand commented:
@PeterKietzmann I think I have addressed most of the issues you've pointed out. For now only the documentation of return codes/errors is on my list.
I've set the authors and copyright in all files to @Nenaco and where applicable myself. I'm not completely sure if this is correct since some of the files might have been based on the at86rf2xx driver and I don't know how this should be handled. If you think that any of those copyright notices do not mention the correct people please let me know.

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_netdev.c:
```diff
@@ -0,0 +1,570 @@
+                    DEBUG("[mrf24j40] TX_CHANNEL_ACCESS_FAILURE\n");
+                }
+                /* check max retries */
+                else if (txstat & 0xc0) {
```
Can `0xc0` be replaced by a more speaking macro?

PeterKietzmann commented on drivers/include/mrf24j40.h:
```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (C) 2017 Neo <neo@nenaco.de>
```
"Neo Nenaco" here and in line 22

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_internal.c:
```diff
@@ -0,0 +1,244 @@
+
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_reg(dev->params.spi, MRF24J40_SHORT_ADDR_TRANS | (addr << MRF24J40_ADDR_OFFSET) | MRF24J40_ACCESS_READ, 0, &value);
```
Would you split this into multiple lines please so we have about 90 symbols per row

PeterKietzmann commented on drivers/mrf24j40/mrf24j40_internal.c:
```diff
@@ -0,0 +1,244 @@
+{
+    spi_acquire(dev->params.spi);
+    gpio_clear(dev->params.cs_pin);
+    spi_transfer_reg(dev->params.spi, MRF24J40_SHORT_ADDR_TRANS | (addr << MRF24J40_ADDR_OFFSET) | MRF24J40_ACCESS_WRITE, value, 0);
```
Split

PeterKietzmann commented:
@bergzand thanks again for taking this up. The code looks great besides my last minor comments. I will give it a try again now. As already stated, this SPI based driver will conflict with #4780. The plan is to merge the SPI rework tomorrow and it should be merged before your PR . Did you already prepare for the SPI rework? Otherwise I can do so and PR against your branch then.

PeterKietzmann commented:
Hmmm.. I'm on stm32f4discovery. Now `gnrc_networking´ hangs after/during initialisation of the device. Any ideas @bergzand?

PeterKietzmann commented:
Still working like charm.

bergzand commented:
@PeterKietzmann The comments should be fixed. I've tried to replace the fixed constants in the code as much as possible. I would be really glad if you have time to implement the SPI rework. I haven't found time for it yet because my own setup decided to quit working this evening. I have time to work on this tomorrow, so if you're too busy, I can do it myself.

bergzand commented:
@PeterKietzmann The SPI API work is done :-) . I've tested it based on the branch of @haukepetersen and everything looked like it was working with gnrc_networking. The work is based on the at86rf2xx SPI API modification.
I've had some problems with `spi_init()` being called too early, but I suspect that's due to my own hacky board setup and not because of RIOT.

The build obviously fails now, but as soon as #4780 is merged I will rebase and push again.
Please let me know whether I missed something essential blocking the merge of this PR

PeterKietzmann commented:
@bergzand that's great. Thanks a lot! Now we 'only' have to wait for SPI but this one is just the next in the queue.

bergzand commented:
Rebased after the merge of #4780

PeterKietzmann commented:
@bergzand wow that was fast :-)! I will give it a last try tomorrow (don't have the hardware with me). @haukepetersen please dismiss your change request. Has been addressed.

bergzand commented:
I noticed the SPI API merge fast due to an email notification, and a rebase+push is not that much work. :-)

I've added a check to the initialization function so that when debugging is enabled in `/drivers/mrf24j40/mrf24j40_internal.c` a register is checked and written to and checked again to verify SPI communication with the mrf24j40. This because we both sometimes seem to have trouble with the SPI communication causing RIOT to hang during initialization without any helpful debug messages.

PeterKietzmann commented:
I can't get it running any more. No time to debug now but it must be that the program hangs before main() starts. Besides of that, please squash you commits to one. The first commit message looks fine.

PeterKietzmann commented:
BTW: It is -as you expected- that the SPI communication during `mrf24j40_init` fails

bergzand commented:
Squashed.

@PeterKietzmann With my current setup (nucleo-f401 with an mrf24j40), I have to call `periph_init()` at the end of the [`board_init()` function in `board.c`](https://github.com/bergzand/RIOT/blob/mrf24j40-nucleo/boards/nucleo-f401/board.c#L37). I'm not sure if this is due to the new SPI API or due to the fact that I power the mrf24j40 via a GPIO pin of the nucleo (Yes, I know, it's ugly, but it works).

PeterKietzmann commented:
That function is already called [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/cpu.c#L55). Tried your last step anyway but w/o success

haukepetersen commented:
by the way, the handling of the `init` functions is quite strange:
- the 'netdev'-init() calls `mrf_setup()`, which calls `mrf_init()`
- the driver offers a set of public, netdev independent functions, but it can not be used stand-alone, as important `init` functionality is tied to the netdev implementation

Seems like this has been copied in parts from the very bad example of the `at86rf` driver?!

haukepetersen commented:
Also the SPI bus is used quite inefficient -> the bus is 'acquired' for every single transfer. Ideally, this should be done only once for each block of transfers, e.g. 'acquire' bus access at the beginning of the initialization or a packet transfer and release it only after the initialization of the packet transfer is complete...

PeterKietzmann commented:
Initial work has been done by students and yes, the atmel driver acted as a start point for that. @bergzand was so nice to take this PR over. However, we can clean up that driver afterwards (if we want to) but in case we can get this running with the SPI rework in the next 1-2 hours I'd vote for taking this PR in.

haukepetersen commented:
Just wanted to state my concerns, I don't see them as blocker for this PR...

Regarding the 'not-working' state, I can't see anything in the code. But I don't have my device here so that I can help debugging. Have you tried running it with different boards?

bergzand commented:
@haukepetersen thank you for your recommendations. I would prefer solving the issues after this PR, unless you consider them blocking for this PR. I would not be suprised if there are some bad practices copied from the at86rf2xx by me or @nenaco. If you could create issues for them and assign them to me,  I can solve them as soon as i've got time for them.

haukepetersen commented:
@bergzand: I don't see them as blocker and they might as well be addressed in follow-up PR(s).

PeterKietzmann commented:
Allright. After I figured out that nucelo-l1 boards are buggy, I can't flash my stm32f0discovery any more, this implementation does not work with stm32f4 and stm32f3 discovery I tried the the samr21-xpro, pulling out the dependency to the `gnrc_netdev_defualt` device. And tada, here it just works fine. So the problem seems not directly related to this driver but to SPI-connected devices with discovery boards. @haukepetersen would you still agree merging this PR?

PeterKietzmann commented:
Ok I will merge this now. Already added it to the list in  #6437. Please help making this driver work with the STM32Fx boards again. Feature freeze will be announces soon. Fixes should be backported to the release branches then. Don't forget!

bergzand commented:
@PeterKietzmann Is it possible for you to check whether this is an issue with the mrf24j40 module or that the SPI module with the STM32Fx boards is broken somehow? I currently run the code from this PR on both a nucleo-f401 (stm32f401RE) and a maple-mini (stm32f103) so I would really like to know if I can reproduce your error. Could this be Nucleo-L1 specific? Furthermore, maybe the problem on the stm32f4 is caused by #6460.

haukepetersen commented:
Actually, it might be, that there is something buggy in the STMs SPI driver (though I wonder how that happened after our extensive testing sessions for the new SPI driver): @kaspar030 was running into trouble today also, when testing the `encx24j60` Ethernet driver with the `nucleo-f334`. Will investigate this tomorrow.

PeterKietzmann commented:
@bergzand do you have a stm32f4discovery board available? I gave it a try again after #6492 was merged but still hangs somewhere during initialization. The radio works as expected with an Atmel samr21-xpro and stand-alone SPI and GPIO look fine on a stm32f4disco board. My configuration is:
```
#ifndef MRF24J40_PARAM_SPI
#define MRF24J40_PARAM_SPI          (SPI_DEV(0))
#endif
#ifndef MRF24J40_PARAM_SPI_CLK
#define MRF24J40_PARAM_SPI_CLK      (SPI_CLK_5MHZ)
#endif
#ifndef MRF24J40_PARAM_CS
#define MRF24J40_PARAM_CS           (GPIO_PIN(0, 4))
#endif
#ifndef MRF24J40_PARAM_INT
#define MRF24J40_PARAM_INT          (GPIO_PIN(4, 15))
#endif
#ifndef MRF24J40_PARAM_RESET
#define MRF24J40_PARAM_RESET        (GPIO_PIN(4, 13))
#endif
```

bergzand commented:
@PeterKietzmann I've borrowed an stm32f4discovery from a friend and I couldn't get it to work either. It seems that it passes the communications check, but hangs when trying a rf state machine. Can you confirm that in your case it also hangs in a loop at https://github.com/RIOT-OS/RIOT/blob/master/drivers/mrf24j40/mrf24j40_getset.c#L476 ?

PeterKietzmann commented:
Confirmed. @bergzand let's continue the discussion in #6519
